File: src/algorithms/python/Longest_Increasing_Subsequence.py
# -*- coding: utf-8 -*-


class Solution(object):
    def lengthOfLIS(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        if nums == []:
            return 0
        n = len(nums)
        dp = [1 for i in range(n)]
        for i in xrange(n):
            maxn = 0
            for j in xrange(i):
                if nums[i] > nums[j] and maxn < dp[j]:
                    maxn = dp[j]
            dp[i] = maxn + 1
        return max(dp)


if __name__ == "__main__":
    solution = Solution()
    print solution.lengthOfLIS([10, 9, 2, 5, 3, 7, 101, 18])

File: src/algorithms/cpp/Longest_Increasing_Subsequence.cpp
#include <iostream>
#include <cstdio>
#include <cstring>
#include <cmath>
#include <cstdlib>
#include <algorithm>
#include <vector>
#include <map>
#include <set>
using namespace std;
#define out(v) cerr << #v << ": " << (v) << endl
#define SZ(v) ((int)(v).size())
const int maxint = -1u>>1;
template <class T> bool get_max(T& a, const T &b) {return b > a? a = b, 1: 0;}
template <class T> bool get_min(T& a, const T &b) {return b < a? a = b, 1: 0;}

class Solution {
public:
    int lengthOfLIS(vector<int>& nums) {
        vector <int> dp (SZ(nums), 1);
        for (int i = 0; i < SZ(nums); ++ i) {
            for (int j = 0; j < i; ++ j) {
                if (nums[j] < nums[i])
                    get_max(dp[i], dp[j] + 1);
            }
        }
        int maxn = 0;
        for (int i = 0; i < SZ(nums); ++ i) {
            get_max(maxn, dp[i]);
        }
        return maxn;
    }
};

int main() {
    return 0;
}
